### ISSUE TITLE
`Index` trait bound makes it impossible to return an owned value

### ISSUE THREAD

**Issue by rocurley** [mitsuhiko/similar#33]
I'm working on json diffing. I have some representation of two jsons in memory, like a `serde_json::Value`. Obviously these cannot be directly diffed, since they're not array-like. So to diff them I need to flatten them into an array of json-lines. I'd rather not actually construct the array, however: I don't want the memory usage. Instead, I'd like to generate the lines on the fly.

I can almost do this by implementing the `Index` trait. It doesn't quite work, however, since I need to return a reference. Ideally, I'd like to be able to pass in an `FnMut(usize) -> T` or something along those lines.

I'm not sure if anybody but me will find this useful, TBH, but I figured there's no harm in asking.

**Comment by mitsuhiko** [comment]
@rocurley Why can you not produce a `&str` (or similar). What the `Index` trait expands to does need to be a reference, but I did not find it massively to be a problem so far. Do you have an example where it does not work for you currently?

**Comment by rocurley** [comment]
So my actual use case is diffing jsons. But you don't need the full complexity of json here to have this issue. Say you want to diff `Vec<Vec<u8>>`s. Now in principle you could just pass that in directly and be fine. But the diff will then consider the inner `Vec<u8>`s to be the unit of diffing, so the diffs will be too coarse. `[[1,2]] -> [[1,2,3]]` shouldn't say the entire input has changed, for example.

So one way to do this is to flatten the input into a vector representation that allows you to reconstruct the original input:
```rust
enum DiffElem {
    VecStart,
    Elem(u8),
}

fn to_diffable(xss : &Vec<Vec<u8>>) -> Vec<DiffElem> {
    let mut out : Vec<DiffElem> = Vec::new();
    for xs in xss {
        out.push(DiffElem::VecStart);
        out.extend(xs.iter().map(|x| DiffElem::Elem(*x)));
    }
    out
}
```
Now this will work, but requires making a copy of the entire input. But it doesn't seem strictly necessary. You could "index" into the diffable representation without ever constructing it:
```rust
// This implementation is ineffecient. To get a more effecient one, we'd
// annotate each subvector with its the index of its VecStart, and use
// binary search to find the subvector we care about.
fn index_diffable(mut i : usize, xss : &Vec<Vec<u8>>) -> Option<DiffElem> {
    for xs in xss {
        if i == 0 {
            return Some(DiffElem::VecStart);
        }
        if i < xs.len() {
            return Some(DiffElem::Elem(xs[i]));
        }
        i -= xs.len();
    }
    None
}
```
But you can't do this with the `Index` trait.

**Comment by mitsuhiko** [comment]
I agree that this case is not great. Generally I'm not a huge fan of the `Index` trait as base. I have already considered alternatives before but never with too much success.

**Comment by kirawi** [comment]
I've been trying to implement a wrapper type to get [`ropey::RopeSlice`](https://docs.rs/ropey/1.3.2/ropey/struct.RopeSlice.html) to work with `similar` using the [iterators it provides](https://docs.rs/ropey/1.3.2/ropey/iter/index.html), but the problem is that they only return owned values.

**Comment by mitsuhiko** [comment]
The original issue I had with `Index` wasn't even that it returns references but that it can be hard to get the bounds checks reliably eliminated (#19). So originally I was trying to do that.

To address this particular thing with borrows I tried to have a custom trait but making both borrows and owned values work is surprisingly complex and I did not go very far with it because I did not at all like what it did to the API. I'm not sure if there are clever ideas I haven't been thinking of yet that would make both cases work.

**Comment by mitsuhiko** [comment]
@cessen the issue is that this `Index` based API is all over the place. This goes really deep in the library.

**Comment by cessen** [comment]
Ah, got it.  If I get some time and the mood strikes, I might take a crack at that as well.  If we're lucky, even if the API goes deep, it still might not be too hard to separate from the algorithm code itself.  If we're lucky... :-)

**Comment by mitsuhiko** [comment]
The other issue with the callback API is that it can't borrow. So it's not even that the callback based API could be the base of the API.

**Comment by kirawi** [comment]
Perhaps a macro that could create two versions of the function; one that borrows and one that returns an owned value?

**Comment by mitsuhiko** [comment]
So one could move somewhat to improving this by going all in on the `IdentifyDistinct` type. So something like this could be done:

```rust
let old = Rope::from_str("...");
let new = Rope::from_str("...");
let h = IdentifyDistinct::<u32>::new_from_iter(old.lines(), new.lines());
let diff = capture_diff(
    Algorithm::Myers,
    h.old_lookup(),
    h.old_range(),
    h.new_lookup(),
    h.new_range(),
);
```

The `new_from_iter` is easy to add, but the returned ops are still annoying to work with. For instance `iter_changes` again requires an `Index` bound.

**Comment by mitsuhiko** [comment]
So now that there are GATs that issue might be worth toying around with again.
